File: backend/app/tools/compensation.py
from __future__ import annotations

from typing import Any

from app.models import Citation, ToolResult


def evaluate(payload: dict[str, Any]) -> ToolResult:
    evidence = payload.get("evidence", [])
    booking = _latest_booking(evidence)
    maintenance_records = _records_for_tool(evidence, "maintenance.search")
    policy_citations = _policy_citations(evidence)
    has_overnight_hvac = any(
        record.get("issue_type") == "hvac"
        and record.get("severity") in {"high", "critical"}
        and ("overnight" in record.get("summary", "").lower() or record.get("status") == "open")
        for record in maintenance_records
    )
    if booking and has_overnight_hvac:
        amount = round(float(booking.get("total_amount", 0)) * 0.30, 2)
        clause = next((c for c in policy_citations if "§4.2" in c.locator), None)
        citations = [clause] if clause else []
        return ToolResult(
            tool="compensation.evaluate",
            data={
                "eligible": True,
                "amount": amount,
                "policy_clause": "§4.2",
                "rationale": "Corroborated overnight HVAC failure qualifies for 30% of affected stay value.",
            },
            citations=[c for c in citations if c],
        )
    return ToolResult(
        tool="compensation.evaluate",
        data={"eligible": False, "amount": 0, "policy_clause": None, "rationale": "No automatic compensation threshold met."},
        citations=[],
    )


def _latest_booking(evidence: list[dict[str, Any]]) -> dict[str, Any]:
    for result in evidence:
        if result.get("tool") == "bookings.lookup":
            return result.get("data", {}).get("booking") or {}
    return {}


def _records_for_tool(evidence: list[dict[str, Any]], tool_name: str) -> list[dict[str, Any]]:
    for result in evidence:
        if result.get("tool") == tool_name:
            return result.get("data", {}).get("records", [])
    return []


def _policy_citations(evidence: list[dict[str, Any]]) -> list[Citation]:
    citations: list[Citation] = []
    for result in evidence:
        if result.get("tool") == "policy.search":
            citations.extend(Citation.model_validate(citation) for citation in result.get("citations", []))
    return citations

File: pibackend/src/tools/compensation.ts
import type { Citation, ToolPayload, ToolResult } from "../types";

export function evaluate(payload: ToolPayload): ToolResult {
  const booking = latestBooking(payload.evidence);
  const maintenanceRecords = recordsForTool(payload.evidence, "maintenance.search");
  const policyCitations = policyCitationsFor(payload.evidence);
  const hasOvernightHvac = maintenanceRecords.some(
    (record) =>
      record.issue_type === "hvac" &&
      ["high", "critical"].includes(String(record.severity)) &&
      (String(record.summary ?? "").toLowerCase().includes("overnight") || record.status === "open")
  );

  if (booking && hasOvernightHvac) {
    const amount = Math.round(Number(booking.total_amount ?? 0) * 30) / 100;
    const clause = policyCitations.find((citation) => citation.locator.includes("4.2"));
    return {
      tool: "compensation.evaluate",
      data: {
        eligible: true,
        amount,
        policy_clause: "\u00a74.2",
        rationale: "Corroborated overnight HVAC failure qualifies for 30% of affected stay value."
      },
      citations: clause ? [clause] : []
    };
  }

  return {
    tool: "compensation.evaluate",
    data: {
      eligible: false,
      amount: 0,
      policy_clause: null,
      rationale: "No automatic compensation threshold met."
    },
    citations: []
  };
}

function latestBooking(evidence: ToolResult[]): Record<string, unknown> {
  return (evidence.find((result) => result.tool === "bookings.lookup")?.data.booking as Record<string, unknown>) ?? {};
}

function recordsForTool(evidence: ToolResult[], tool: string): Array<Record<string, unknown>> {
  return (evidence.find((result) => result.tool === tool)?.data.records as Array<Record<string, unknown>>) ?? [];
}

function policyCitationsFor(evidence: ToolResult[]): Citation[] {
  return evidence.filter((result) => result.tool === "policy.search").flatMap((result) => result.citations);
}
